src/port1.0/portconfigure.tcl

```diff
@@ -528,98 +599,471 @@ proc portconfigure::configure_get_default_compiler {} {
+        return [expr {${os.major} >= 11}]
+    } else {
+        return yes
+    }
+}
+#
+# internal utility procedure to get a version larger than any compiler version
+proc portconfigure::max_compiler_version {} {
+    return 999999999999999999999.0
+}
+#
+# https://releases.llvm.org/3.1/docs/ClangReleaseNotes.html#cchanges
+# https://gcc.gnu.org/c99status.html
+# https://gcc.gnu.org/wiki/C11Status
+# https://trac.macports.org/wiki/XcodeVersionInfo
+#-------------------------------------------------------------------
+#|  C Standard  |   Clang   |  Xcode Clang  |   Xcode   |    GCC    |
+#|-----------------------------------------------------------------|
+#| 1989 (C89)   |     -     |        -      |     -     |     -     |
+#| 1999 (C99)   |     -     |    211.10.1   |    4.2    |    4.5    |
+#| 2011 (C11)   |    3.1    |    318.0.61   |    4.3    |    4.9    |
+#-------------------------------------------------------------------
+#
+# https://clang.llvm.org/cxx_status.html
+# https://gcc.gnu.org/projects/cxx-status.html
+# https://en.cppreference.com/w/cpp/compiler_support
+# Xcode release notes
+# https://trac.macports.org/wiki/XcodeVersionInfo
+#-------------------------------------------------------------------
+#| C++ Standard |   Clang   |  Xcode Clang  |   Xcode   |    GCC    |
+#|-----------------------------------------------------------------|
+#| 1998 (C++98) |     -     |       -       |     -     |     -     |
+#| 2011 (C++11) |    3.3    |   500.2.75    |    5.0    |   4.8.1   |
+#| 2014 (C++14) |    3.4    |   600.0.54    |    6.1    |     5     |
+#| 2017 (C++17) |    5.0    |   902.0.39.1  |    9.3    |     7     |
+#-------------------------------------------------------------------
+#
+# https://openmp.llvm.org
+# https://gcc.gnu.org/wiki/openmp
+# https://trac.macports.org/wiki/XcodeVersionInfo
+#|---------------------------------------------------------------
+#| OpenMP Version |  Clang  |  Xcode Clang  |  Xcode  |   GCC   |
+#|---------------------------------------------------------------
+#|      2.5       |   3.8   |    Future?    | Future? |   4.2   |
+#|      3.0       |   3.8   |    Future?    | Future? |   4.4   |
+#|      3.1       |   3.8   |    Future?    | Future? |   4.7   |
+#|      4.0       | Partial |    Future?    | Future? |   4.9   |
+#|      4.5       | Partial |    Future?    | Future? |   ???   |
+#----------------------------------------------------------------
+#
+# utility procedure: get minimum command line compilers version based on restrictions
+proc portconfigure::get_min_command_line {compiler} {
+    global compiler.c_standard compiler.cxx_standard compiler.openmp_version compiler.thread_local_storage
+    set min_values 1.0
+    switch ${compiler} {
+        clang {
+            if {${compiler.c_standard} >= 2011} {
+                lappend min_values 318.0.61
+            } elseif {${compiler.c_standard} >= 1999} {
+                lappend min_values 211.10.1
```

Comment:
The choice of C99 was made in [a previous comment](https://github.com/macports/macports-base/pull/88#discussion_r193841635).

Reply:
Clicking that link doesn't seem to show the comment.

Reply:
Sorry about that.
It takes a couple of seconds to load, but the [link](https://github.com/macports/macports-base/pull/88#discussion_r193841635) works for me.

It was a comment by @jeremyhu on June 7th.
I am afraid I do not know how else to link to the comment.

I do not have strong feelings about the default value of `compiler.c_standard`.
I was put in what others (more experienced than I) recommended.

Reply:
OK, I see the one you mean. It's not clear whether Jeremy understood that this was excluding compilers without full C99 support, rather than just passing `-std=c99` to the compiler. I think perhaps those should be two different settings. Perhaps extra values like `1999-strict` for compiler.c_standard? Or another option like `compiler.strict_standards_compliance`?

Also note that we still don't technically have full C99 compliance even with the recent enough compilers because we're not using `-pedantic-errors` (and `-fextended-identifiers` with gcc < 5).

Reply:
It just occurs to me that this might be a somewhat moot question.
The default value for `compiler.cxx_standard` is 2011.
This was done to [reduce the likelihood of crossing boundaries between C++ libraries](https://github.com/macports/macports-base/pull/88#discussion_r193839696).

This means that without modification, ports will default to GCC >= 4.8.1, Xcode >= 5, or Clang >= 3.3 anyway.
Of course, any port is free to override this.

Reply:
You're right, that's far worse. Adding these dependencies to every port regardless of whether they even uses a C++ compiler? Nobody has the time to go through and turn it off for all the ports that don't need it.

Reply:
So are you suggesting that `compiler.c_standard` and `compiler.cxx_standard` should default to `1989` and `1998` respectively?
If so, do you have any thoughts on how to solve the mixing of C++ libraries issue?

I do not have strong feelings about the default values but was just trying to address the concerns of others.

Reply:
This is only really an issue on Tiger and Leopard if we're going to flip the default stdlib to libc++ on 10.6+. What happens if you ask for +universal on Leopard, for example? (Although there are probably users who want to build 3-way universal on 10.6 too…)

I think the safest way to proceed is to provide these mechanisms but not change the default policies yet. If mixing of stdlibs is a serious issue, it's probably better solved by rev-upgrade and something similar to what we do for architectures when calculating dependencies. Don't forget it goes both ways: if it's a problem for only some ports to use macports-libstdc++, then it's also a problem for only some ports to use the system libstdc++.

Reply:
Forgive me, @jmroot, but I do not quite understand your last point.
Sorry if I'm missing something.

>  if it's a problem for only some ports to use macports-libstdc++, then it's also a problem for only some ports to use the system libstdc++.

Isn't that an argument to default `compiler.cxx_standard` to `2011`, thereby ensuring that **all** ports use the same standard library?

> This is only really an issue on Tiger and Leopard if we're going to flip the default stdlib to libc++ on 10.6+.

I have not been following this work very carefully.
Will it still be possible for users to keep the standard library as libstdc++?
If so, aren't we in the same position we are now, trying to ensure consistent standard libraries?

> it's probably better solved by rev-upgrade

A fair point, but wouldn't that mean we would continually be having to add `compiler.cxx_standard 2011` to ports that don't require C++11 just because they link against others that do?

Reply:
Sorry for the delay in replying.

> Forgive me, @jmroot, but I do not quite understand your last point.
> Sorry if I'm missing something.
> 
> > if it's a problem for only some ports to use macports-libstdc++, then it's also a problem for only some ports to use the system libstdc++.
> 
> Isn't that an argument to default `compiler.cxx_standard` to `2011`, thereby ensuring that **all** ports use the same standard library?

Only if you prevent individual ports from changing it.

> > This is only really an issue on Tiger and Leopard if we're going to flip the default stdlib to libc++ on 10.6+.
> 
> I have not been following this work very carefully.
> Will it still be possible for users to keep the standard library as libstdc++?
> If so, aren't we in the same position we are now, trying to ensure consistent standard libraries?

They will be able to change it in macports.conf if they really want to, but it will be an unsupported configuration like libc++ on older systems is now.

> > it's probably better solved by rev-upgrade
> 
> A fair point, but wouldn't that mean we would continually be having to add `compiler.cxx_standard 2011` to ports that don't require C++11 just because they link against others that do?

Is that really needed in most cases? Even if so, I think a mechanism that ensures a consistent stdlib+standard in all dependencies should be possible.

Reply:
Since I never use the older systems myself, I am more than happy to implement whatever others decide on the default C and C++ standards.
However, consensus would seem to be elusive on this issue.
Perhaps one of the following could be the next concrete step to move this forward:
1. Merge the pull request for more widespread testing and to
   * see if the changes cause more problems than they fix
   * see how problematic [lack of caching](https://github.com/macports/macports-base/pull/88#issuecomment-435279559) is,
1. Take a vote (perhaps through the mailing list) on the default standards, or
1. Let those who regularly use and support older systems decide.

It seems to me that (1) is the best choice, so we can have better data on which to make decisions.
I have run tests, but my testing infrastructure is very limited.

Reply:
> Is that really needed in most cases? Even if so, I think a mechanism that ensures a consistent stdlib+standard in all dependencies should be possible

The system libraries are all linked against /usr/lib/libstdc++.dylib on those systems, so I don't see how we can ever get away from the mixture of stdlibs. I have not tried dropping a symlink from the newer libstdc++.dylib from gcc7 into /usr/lib/ to see what would happen. It actually might work -- it's supposed to be backward compatible -- but the system would be unusable if it didn't, or during some gcc7 upgrade I guess.

So far (two years in) I am not aware of any cases where there has been an issue with our current stdlib setup (other than the gcc headers being different enough from the libc++ headers to be a PITA sometimes). 

There was a theoretical issue wherein an object might be created by a code bit linked to one stdlib and freed by a code bit linked against the other stdlib, which would apparently cause an error if it happened, but I have never seen it happen so far.

Those systems still need apple-gcc42 for many things. Changing all ports to build against /opt/local/lib/libstdc++.dylib would most likely mean changing the build of apple-gcc42 to link against that stdlib, and then forcing apple-gcc42 to be used for port building instead of /usr/bin/gcc-4.2 or llvm-gcc42. I don't know if that is possible. Perhaps it is.

I am not seeing this as a huge issue to fix at the moment. The status quo seems to be making everyone happy.

Reply:
This pull request now defaults the C standard to 1989 and the C++ standard to 1998.